rs/sns/governance/token_valuation/src/lib.rs

```diff
@@ -293,73 +294,130 @@ impl IcpsPerTokenClient for IcpsPerIcpClient {
 
 struct IcpsPerSnsTokenClient<MyRuntime: Runtime + Send + Sync> {
     swap_canister_id: CanisterId,
+    sns_token_ledger_canister_id: CanisterId,
     _runtime: PhantomData<MyRuntime>,
 }
 
 impl<MyRuntime: Runtime + Send + Sync> IcpsPerSnsTokenClient<MyRuntime> {
-    pub fn new(swap_canister_id: CanisterId) -> Self {
+    pub fn new(
+        swap_canister_id: CanisterId,
+        // This is used to determine the amount of inflation since genesis/swap execution.
+        sns_token_ledger_canister_id: CanisterId,
+    ) -> Self {
         Self {
             swap_canister_id,
+            sns_token_ledger_canister_id,
             _runtime: Default::default(),
         }
     }
 
     async fn fetch_icps_per_sns_token(&self) -> Result<Decimal, ValuationError> {
-        // Fetch SNS token price from swap.
-        let get_derived_state_response = self.call(GetDerivedStateRequest {}).await?;
-
-        // Read the relevant field(s) out of the responses. Here, a floating point field is used. In
-        // general, floating point should not be used for financial accounting, but it is ok here,
-        // because we are using this to come up with a valuation, and valuations are not super
-        // precise in the same way that (for example) a bank account balance is supposed to be.
-        let sns_tokens_per_icp: f64 =
-            get_derived_state_response
-                .sns_tokens_per_icp
-                .ok_or_else(|| {
-                    ValuationError::new_mismatch(format!(
-                        "Response from swap ({}) get_derived_state call did not \
-                         contain sns_tokens_per_icp: {:#?}",
-                        self.swap_canister_id, get_derived_state_response,
-                    ))
-                })?;
+        // (Concurrently) fetch the various pieces that we need to sythensize the result:
+        let (get_derived_state_result, initial_supply_e8s_result, current_supply_result) = join!(
+            // 1. SNS token price from swap.
+            call::<_, MyRuntime>(self.swap_canister_id, GetDerivedStateRequest {}),
```

Comment:
why call here, and call_with_cleanup a bit lower?

Reply:
To answer your question, call does not work when a method takes 0 arguments. Rather, it assumes the method is of the form

```
foo : (A) -> (B);
```

Maybe I should just get rid of call. Is that what you are suggesting? I guess you are leaving this up to me, since you approved, and this feedback seems to be "just" a question.

(call is nice, because you do not need to specify the return type. It deduces that based on the Request trait.)